### PR TITLE
chore(agent): separate PR policy from background watchers

### PR DESCRIPTION
**Pull request by cyhhao**
## What changed

This follow-up establishes the skill responsibility boundary after Avibe PR #1213 merged.

- Keep pr-delivery-loop as the Avibe project policy for branch scope, contracts, Codex review, CI, thread resolution, circuit breaking, and close-out.
- Delegate all managed waiting to the background-watch-hook skill by name.
- Remove the duplicate PR and Actions waiters and their tests from pr-delivery-loop.
- Update Avibe AGENTS guidance and add a contract test that prevents the duplicate waiter bundle from returning.

## Evidence

- Unit: 100 focused Harness, GitHub PR waiter, and GitHub Actions waiter tests passed.
- Contract: the skill guidance test verifies the dependency boundary and the canonical background-watch implementation.
- Scenario: not applicable; this change is workflow documentation and test coverage.
- Residual manual checks: inspect the rendered Skill and confirm the one-watch workflow after the PR review loop is armed.

## Dependencies

- None. This is the canonical Avibe cleanup; the other repository copies remain closed until this policy passes review.

## Known by design

- background-watch-hook source and behavior are unchanged in this PR.
- Companion migration PRs were closed as superseded; synchronization will follow this canonical review.

### CODE REVIEW

**Comment by cyhhao** [comment]
@codex review

**Comment by cyhhao** [comment]
@codex review

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 1]
```diff
@@ -180,97 +184,35 @@ turn ends because you armed a watch and are waiting, say exactly that.
...
-  `endCursor` and fetches complete mutable REST collections rather than `since`
-  slices. The sole wake condition is inequality between the current normalized
-  snapshot and the committed snapshot. Per-field checks only describe that
-  difference, so additions, edits, removals, resolution changes, reactions,
-  status changes, and head changes cannot be silently omitted by an incomplete
-  trigger list.
+- Follow `background-watch-hook` for waiter commands, state, catch-up,
+  filtering, settling, retries, and delivery acknowledgement. Those mechanics
+  belong to the reusable skill; this policy only constrains when and why the
+  watch is armed.
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md]: Fixed in 1dc4405d: the project policy now requires catch-up mode or a pre-seeded owner-specific baseline when a watch is armed after a push or review trigger. The generic skill documents catch-up for this case.

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 2]
```diff
@@ -180,97 +184,35 @@ turn ends because you armed a watch and are waiting, say exactly that.
...
-  GitHub viewer login. It must resolve that identity before polling; if `/user`
-  is unavailable for `--pr`, fail closed or explicitly pass
-  `--include-self-comments` and account for those events in the cursor plan.
-  `--new-prs` does not require a viewer lookup because it has no PR-local
-  self-comment filter.
+- Use `background-watch-hook` to create a one-shot PR watch for each review
+  phase; never use `--forever` for a delivery loop. Re-arm after every round.
+  When CI is the only remaining gate, replace the PR-activity watch with an
+  exact-head Actions watch because PR activity cannot wake on a check-only
+  transition.
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:191]: Fixed in 1dc4405d: the Actions waiter retains every distinct matching run ID at the target SHA, so an older pending or failed rerun cannot be hidden by a newer success. Added regression coverage.

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 3]
```diff
@@ -180,97 +184,35 @@ turn ends because you armed a watch and are waiting, say exactly that.
...
-- The PR waiter filters your own reviews and comments by the authenticated
-  GitHub viewer login. It must resolve that identity before polling; if `/user`
-  is unavailable for `--pr`, fail closed or explicitly pass
-  `--include-self-comments` and account for those events in the cursor plan.
-  `--new-prs` does not require a viewer lookup because it has no PR-local
-  self-comment filter.
+- Use `background-watch-hook` to create a one-shot PR watch for each review
+  phase; never use `--forever` for a delivery loop. Re-arm after every round.
+  When CI is the only remaining gate, replace the PR-activity watch with an
+  exact-head Actions watch because PR activity cannot wake on a check-only
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:190]: Fixed in 1dc4405d: the PR waiter persists the last observed head SHA and emits a head-change activity event, so a push wakes the review loop even without a new comment.

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 4]
```diff
@@ -283,15 +225,12 @@ turn ends because you armed a watch and are waiting, say exactly that.
...
-  still unchanged. If prior-review completion cannot be established, require
-  the bot-authored pass comment instead of accepting a reaction-only pass. The
-  waiter's `pr_reaction` output is the durable evidence for this phase, so it
-  must never be omitted by `--event-limit` even in a large review batch.
+  new head, establish the current activity baseline through
+  `background-watch-hook`, and do not push another head while the new review is
+  pending. Accept the reaction only when that watch reports it as new for the
+  current review phase, the prior-head review was already terminal, and the PR
+  head is unchanged. If those facts cannot be established, require the
+  bot-authored exact-head pass comment instead.
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:239]: Fixed in 1dc4405d: Codex pass reactions accept both API login forms, with regression coverage for the unsuffixed login.

[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:239]: Fixed in 1dc4405d: Codex pass reactions are appended outside the normal event limit, so required pass evidence cannot be truncated while the reaction cursor advances.

File: .agents/skills/pr-delivery-loop/SKILL.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md]: Fixed in 1dc4405d: review and issue comments now persist mutable-content fingerprints and edited existing items are reported again. Legacy numeric-only cursors seed their first fingerprint silently.

[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md]: Fixed in 1dc4405d: both waiters now exit immediately for terminal GitHub HTTP failures, while retryable HTTP and network failures retain retryable exit behavior.

[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md]: Fixed in 1dc4405d: with authenticated self-comment filtering enabled, an unresolved GitHub viewer login is terminal instead of silently disabling the filter. Explicit include-self-comments remains the opt-in escape hatch.

File: .agents/skills/pr-delivery-loop/SKILL.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:191]: Fixed in 1dc4405d: the merge-gate Actions example omits only-on-failure so a green exact-head result wakes the final gate; that flag remains documented only for intentionally silent green watches.

**Comment by cyhhao** [comment]
@codex review

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 5]
```diff
@@ -10,19 +10,24 @@ description: The implementation-lane standard for delivering a PR across the Avi
...
 - In the multi-repo workspace, the project-level skill entry is a symlink to
   the Avibe copy. Update the canonical copy first, then sync every repository
-  copy in the same change.
+  copy only after the canonical Avibe change passes review.
+
+## Dependency boundary
+
+- Use the `background-watch-hook` skill for every managed wait. It owns the
+  reusable `vibe watch` workflow and the GitHub PR, issue, and Actions waiter
+  implementations.
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:23]: Fixed in c12572e8: managed pending transactions now persist the rendered report alongside staged cursors and replay it before authentication or any GitHub request when the prior delivery was not acknowledged.

File: skills/background-watch-hook/SKILL.md [hunk 6]
```diff
@@ -260,7 +267,9 @@ The login is reused only while the token still fingerprints to the account it wa
 resolved for, and an explicit `--since-review-comment-id` /
 `--since-issue-comment-id` drops the matching saved `since` so the replay it asks
 for is not narrowed away. It is written on every cursor advance and on timeout,
-replaced atomically. Cursors that cover a reported event are not committed by the
+replaced atomically. The state also records the last observed PR head and mutable
+review/comment fingerprints, so a pushed head or an edited existing comment is
+activity even when no new numeric ID exists. Cursors that cover a reported event are not committed by the
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/SKILL.md]: Fixed in c12572e8: reviews now have their own persisted mutable fingerprints and normalized snapshot entries, so body, state, or commit changes on an existing review ID wake the waiter. Regression coverage includes an edited review.

File: skills/background-watch-hook/SKILL.md [hunk 7]
```diff
@@ -195,9 +195,16 @@ vibe watch add \
...
     --actionable-only \
     --settle 20 \
+    --catch-up \
     --interval 60
 ```
 
+Use `--catch-up` when the watch is armed after the triggering push or review
+comment. Without an explicit state baseline, the first poll treats activity
+already present as history and can wait past the event that should wake the
+Agent. A pre-seeded owner-specific `--state-file` is the alternative when the
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/SKILL.md]: Fixed in c12572e8: review-loop guidance now requires a complete owner-specific baseline seeded before the push or review trigger. Catch-up is documented only for intentional historical replay, and seed mode snapshots existing threads without reporting them.

File: .agents/skills/pr-delivery-loop/SKILL.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:23]: Fixed in c12572e8: the shared GitHub helper now follows GitHub CLI precedence by selecting GH_TOKEN before GITHUB_TOKEN, with regression coverage.

[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:23]: Fixed in c12572e8: the waiter validates the remote PR before creating or adopting the state file. Invalid repository or PR targets leave no state-file claim, with regression coverage.

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 8]
```diff
@@ -180,97 +184,39 @@ turn ends because you armed a watch and are waiting, say exactly that.
...
-  identical on a dashboard, so a brief must never ask a lane to "self-certify
-  enabled" — that phrase prescribes exactly the unsound check.
-- Preserve the watch ID when re-arming: update the existing definition when its
-  command or message changed, then resume that same ID so its owner stamp remains
-  compatible with the persisted state file and the one-shot is enabled. If an
-  ownership handoff is unavoidable, rotate to a fresh state path and seed it
-  before starting the replacement. A watch you delete on purpose is not a
-  liveness break while you are mid-run: the running turn *is* the liveness.
-  Re-arm after the push, then verify both ways.
+  the real next review lands with nobody watching. End every round by using the
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:205]: Fixed in c12572e8: the canonical waiter paginates every GraphQL reviewThread, persists resolution states, and wakes on resolve, reopen, or removal transitions. The complete mutable PR snapshot is covered by focused tests.

File: .agents/skills/pr-delivery-loop/SKILL.md [hunk 9]
```diff
@@ -10,19 +10,24 @@ description: The implementation-lane standard for delivering a PR across the Avi
...
+- Use the `background-watch-hook` skill for every managed wait. It owns the
+  reusable `vibe watch` workflow and the GitHub PR, issue, and Actions waiter
+  implementations.
+- This skill owns Avibe-specific delivery policy: branch and scope rules,
+  review and CI gates, thread resolution, circuit breaking, authority, and
+  close-out criteria.
+- Refer to the dependency by skill name. Do not hard-code an installation path,
+  vendor its scripts here, or hand-roll a replacement waiter. If
+  `background-watch-hook` is unavailable, install or enable it before starting
+  a review loop, or report the missing dependency as a blocker.
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply .agents/skills/pr-delivery-loop/SKILL.md:30]: Fixed in c12572e8: the generic skill now shows a portable active-skill resolver and all examples invoke waiters through that resolved directory. The project policy continues to reference only the background-watch-hook skill boundary.

**Comment by cyhhao** [comment]
@codex review

File: skills/background-watch-hook/SKILL.md [hunk 10]
```diff
@@ -190,14 +206,29 @@ vibe watch add \
...
     --repo avibe-bot/avibe \
     --pr 151 \
     --actionable-only \
     --settle 20 \
     --interval 60
 ```
 
+Before a push or review trigger, seed an owner-specific state file from the
+current complete PR snapshot. Arm the post-action watch with that same file so
+activity that lands during the handoff remains visible:
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/SKILL.md]: Fixed in 67527165. Every managed PR example now declares an owner-specific state file and passes that exact file to wait_pr.py. Future-event examples seed the complete baseline before the watched action; catch-up uses its own durable file without seeding by design.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 11]
```diff
@@ -1610,10 +2185,13 @@ def _report_new_pr(output: str, previous: int) -> None:
                 )
                 return 1
             print(f"GitHub API error during polling: {err.code} {err.reason}", file=sys.stderr)
-            continue
+            return RETRY_EXIT_CODE if is_retryable_http_error(err) else 1
+        except urllib.error.URLError as err:
+            print(f"GitHub network error during polling: {err.reason}", file=sys.stderr)
+            return RETRY_EXIT_CODE
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py]: Fixed in 67527165 as a waiter-level contract, not a supervisor workaround. PR and Actions waiters now classify GitHub failures centrally, retry startup failures with bounded backoff, and keep the same one-shot process alive across retryable polling failures. Regression tests cover transient startup recovery, retry exhaustion, terminal failures, and mid-poll recovery.

File: skills/background-watch-hook/SKILL.md [hunk 12]
```diff
@@ -179,6 +179,22 @@ Use bundled waiters as examples or as ready-to-run building blocks. The main ski
...
+The waiter is distributed with this skill, not with the caller's repository.
+Resolve the active skill directory before constructing a watch command:
+
+```bash
+BACKGROUND_WATCH_HOOK_SKILL_FILE="${BACKGROUND_WATCH_HOOK_SKILL_FILE:-}"
+if [ -z "$BACKGROUND_WATCH_HOOK_SKILL_FILE" ]; then
+  for SKILL_ROOT in "${CODEX_HOME:-$HOME/.codex}/skills" "${AGENTS_HOME:-$HOME/.agents}/skills"; do
+    [ -d "$SKILL_ROOT" ] || continue
+    BACKGROUND_WATCH_HOOK_SKILL_FILE="$(find "$SKILL_ROOT" -path '*/background-watch-hook/SKILL.md' -print -quit)"
+    [ -n "$BACKGROUND_WATCH_HOOK_SKILL_FILE" ] && break
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/SKILL.md:208]: Fixed in 67527165. The documented resolver now checks the current repository root for skills/background-watch-hook and .agents/skills/background-watch-hook before falling back to installed Codex/Agents roots. The focused guidance test and a shell smoke check cover a fresh Avibe checkout.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 13]
```diff
@@ -1231,39 +1670,32 @@ def main() -> int:
         if token_fingerprint is not None and _saved_str(saved, "token_fingerprint") == token_fingerprint:
             viewer_login = _saved_str(saved, "viewer_login")
         viewer_login = viewer_login or get_authenticated_login(token)
+        if token is not None and viewer_login is None:
+            print(
+                "Could not resolve the authenticated GitHub login; refusing to poll while self-comment filtering is enabled.",
+                file=sys.stderr,
+            )
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py]: Fixed in 67527165. Viewer identity is now resolved only for single-PR mode, where self-comment filtering needs it. Repository-wide new-PR mode no longer calls /user, with an app-token regression test that fails if the lookup occurs.

**Comment by cyhhao** [comment]
@codex review

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 14]
```diff
@@ -1191,13 +1661,82 @@ def main() -> int:
...
+    token_fingerprint = _token_fingerprint(token)
+    viewer_login = None
+    if args.pr is not None and not args.include_self_comments:
+        # The stored login spares a /user request on every cycle of a forever watch,
+        # but only while the token still belongs to the account it was resolved for.
+        # A rotated or swapped credential would otherwise keep filtering out the old
+        # account's comments and let the new account's own comments wake the Agent.
+        if token_fingerprint is not None and _saved_str(saved, "token_fingerprint") == token_fingerprint:
+            viewer_login = _saved_str(saved, "viewer_login")
+        viewer_login = viewer_login or get_authenticated_login(token)
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py]: Fixed in bb1e4a64. The authenticated viewer lookup now uses the same classified, bounded initial-request retry path as PR-state bootstrap. Transient network and retryable HTTP failures are retried, while terminal and unusable responses still fail closed. Covered by strict helper and transient-retry regressions.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 15]
```diff
@@ -385,29 +701,66 @@ def _visible(items: list[dict[str, Any]]) -> list[dict[str, Any]]:
...
-    next_review_comment_cursor = max(review_comment_cursor, max_id(new_review_comments))
-    next_issue_comment_cursor = max(issue_comment_cursor, max_id(new_issue_comments))
+    has_head_event = bool(previous_head_sha and current_head_sha and previous_head_sha != current_head_sha)
+    raw_threads = state.get("review_threads")
+    current_thread_states = (
+        _review_thread_state_map(raw_threads if isinstance(raw_threads, list) else [])
+        if review_threads_available
+        else dict(review_thread_states or {})
+    )
+    thread_changes = _review_thread_state_changes(current_thread_states, review_thread_states or {})
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py:709]: Fixed in bb1e4a64. A fresh normal PR wait now seeds head identity, mutable-object fingerprints, every review-thread state, and the normalized snapshot before its first render. Existing threads are therefore baseline state rather than absent-to-state events. Added a fresh-watch regression with an existing unresolved thread.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 16]
```diff
@@ -1467,15 +2009,19 @@ def _settle(
...
                     )
+                except urllib.error.HTTPError as err:
+                    print(f"Settle GitHub API error: {err.code} {err.reason}", file=sys.stderr)
+                    if not is_retryable_http_error(err):
+                        raise
+                    return best
+                except urllib.error.URLError as err:
+                    print(f"Settle GitHub network error: {err.reason}", file=sys.stderr)
                     return best
+                except Exception:
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py]: Fixed in bb1e4a64. Settle re-polls now pass through the shared GitHub request taxonomy, and any enrichment failure returns the already detected best batch instead of suppressing it. Added coverage for TimeoutError, ConnectionError, OSError, and the existing URL error case.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 17]
```diff
@@ -261,11 +348,202 @@ def _format_pull_request(pr: dict[str, Any]) -> str:
...
+    review_comments = _visible(state["review_comments"])
+    issue_comments = _visible(state["issue_comments"])
+    reactions = [reaction for reaction in state["reactions"] if _is_codex_pass_reaction(reaction)]
+    normalized_reactions = _normalized_item_map(reactions, fields=("content",))
+    for reaction in normalized_reactions.values():
+        reaction["author"] = "chatgpt-codex-connector"
+
+    if review_threads_available:
+        raw_threads = state.get("review_threads")
+        review_threads = _review_thread_state_map(raw_threads if isinstance(raw_threads, list) else [])
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py:469]: Addressed in bb1e4a64 by making the event contract explicit and consistent. Author and pattern filters suppress review/comment payloads; review-thread status is a separate mutable signal because another actor may later resolve or reopen it. The CLI help, generic skill documentation, normalized snapshot behavior, and a cross-author regression now all enforce that contract.

File: skills/background-watch-hook/scripts/wait_pr.py [hunk 18]
```diff
@@ -1212,41 +1751,62 @@ def main() -> int:
...
-    # baseline to be derived from a `since`-narrowed fetch, which no longer
-    # contains the PR's full history.
     resume_cursors = {key: _saved_int(saved, key) for key in STATE_CURSOR_KEYS}
     resumed = not args.catch_up and all(value is not None for value in resume_cursors.values())
-    # An explicit --since-*-comment-id asks for a replay from that id. The saved
-    # `since` timestamp is only ever a shortcut for the saved cursor, so keeping it
-    # would narrow the fetch to comments newer than the last poll and hide exactly
-    # the history the flag asked to see.
+    missing_baselines = _missing_pr_baselines(saved) if resumed and args.pr is not None else []
+    if missing_baselines and two_phase:
```
[the remark this answers is not in the document]
Comment (cyhhao) [reply skills/background-watch-hook/scripts/wait_pr.py]: Fixed in bb1e4a64. Incomplete resumed PR baselines are now rejected for both manual and managed runs. Only an explicit cursor replay or deliberate catch-up may bypass the complete-snapshot requirement; otherwise the operator must reseed. Added a manual legacy-state regression that proves the file is left untouched.

**Comment by cyhhao** [comment]
@codex review
